File: server/hardware_ai_test.go
package main

import (
	"math"
	"testing"
)

func hardwareFixture(x, y int) (State, [18]int) {
	const u = 22.4 / 576
	s := initial()
	var distances [18]int
	for i := range s.Players {
		s.Players[i].Stun = 100
		distances[i] = 1000
	}
	p := &s.Players[1]
	p.X, p.Z, p.Stun, p.aiWait = float64(576-y)*u, float64(x-320)*u, 0, 0
	s.Ball.X, s.Ball.Z, s.Ball.Owner = p.X, p.Z, 1
	return s, distances
}
func TestHardwareThrowPriority(t *testing.T) {
	const u = 22.4 / 576
	s, d := hardwareFixture(56, 700)
	if p := s.hardwareThrow(1, 255, &d); p == nil || p.high {
		t.Fatal("multiplier")
	}
	s.Multiplier = 2
	if s.hardwareThrow(1, 255, &d) != nil {
		t.Fatal("maximum multiplier")
	}
	s.Multiplier = 0
	s.Players[16].X, s.Players[16].Z, s.Players[16].Stun = (576-650)*u, (56-320)*u, 0
	if s.hardwareThrow(1, 255, &d) != nil {
		t.Fatal("blocked lane")
	}
	s.Ball.Charged = true
	if p := s.hardwareThrow(1, 255, &d); p == nil || p.key != 3 {
		t.Fatal("electroball lane")
	}
}
func TestHardwareWallDirection(t *testing.T) {
	const u = 22.4 / 576
	s, d := hardwareFixture(160, 600)
	p := s.hardwareThrow(1, 255, &d)
	if p == nil || !p.high || p.key != 2 || math.Abs(p.x-(576-464)*u) > 1e-12 || math.Abs(p.z-(32-320)*u) > 1e-12 {
		t.Fatal("wall diagonal", p)
	}
	s.Period = 2
	if s.hardwareThrow(1, 255, &d) != nil {
		t.Fatal("backwards wall throw")
	}
	s.Players[1].X, s.Players[1].Z = (576-552)*u, (480-320)*u
	if p := s.hardwareThrow(1, 255, &d); p == nil || !p.high || p.key != -2 {
		t.Fatal("mirrored diagonal", p)
	}
}
func TestHardwareZapper(t *testing.T) {
	const u = 22.4 / 576
	s, d := hardwareFixture(100, 900)
	if p := s.hardwareThrow(1, 49, &d); p == nil || p.high || p.key != -1 {
		t.Fatal("zapper", p)
	}
	if s.hardwareThrow(1, 50, &d) != nil {
		t.Fatal("strict aggression")
	}
	s.Ball.Charged = true
	if s.hardwareThrow(1, 49, &d) != nil {
		t.Fatal("already charged")
	}
	s.Ball.Charged = false
	s.Players[1].X, s.Players[1].Z = (576-986)*u, (126-320)*u
	if p := s.hardwareThrow(1, 49, &d); p == nil || p.high {
		t.Fatal("inclusive rectangle", p)
	}
	s.Players[1].Z = (127 - 320) * u
	if s.hardwareThrow(1, 49, &d) != nil {
		t.Fatal("outside rectangle")
	}
}
func TestHardwareStartsThrow(t *testing.T) {
	s, _ := hardwareFixture(56, 700)
	s.simulate(simulationStep, [2]Input{}, [2]bool{})
	p := s.Players[1]
	if p.throwMode != 2 || p.FX != 1 || p.FZ != 0 {
		t.Fatal("hardware did not override legacy trigger", p)
	}
}

File: src/hardware-ai.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, step, simulationStep } from "./game.js";
import { hardwareThrow } from "./hardware-ai.js";
const u = 22.4 / 576;
function fixture(x, y) {
  const s = initial();
  for (const p of s.players) p.stun = 100;
  Object.assign(s.players[1], {
    x: (576 - y) * u,
    z: (x - 320) * u,
    stun: 0,
    aiWait: 0,
  });
  Object.assign(s.ball, { x: s.players[1].x, z: s.players[1].z, owner: 1 });
  return s;
}
const distances = Array(18).fill(1000);
test("hardware priority: multiplier low throw, maximum multiplier and blocked/electric lane", () => {
  const s = fixture(56, 700);
  assert.equal(hardwareThrow(s, 1, 255, distances).high, false);
  s.multiplier = 2;
  assert.equal(hardwareThrow(s, 1, 255, distances), null);
  s.multiplier = 0;
  Object.assign(s.players[16], {
    x: (576 - 650) * u,
    z: (56 - 320) * u,
    stun: 0,
  });
  assert.equal(hardwareThrow(s, 1, 255, distances), null);
  s.ball.charged = true;
  assert.equal(hardwareThrow(s, 1, 255, distances).key, 3);
});
test("high wall throws require an attacking diagonal and mirror after halftime", () => {
  const s = fixture(160, 600);
  let p = hardwareThrow(s, 1, 255, distances);
  assert.equal(p.high, true);
  assert.equal(p.key, 2);
  assert.equal(p.x, (576 - 464) * u);
  assert.equal(p.z, (32 - 320) * u);
  s.period = 2;
  assert.equal(hardwareThrow(s, 1, 255, distances), null);
  Object.assign(s.players[1], { x: (576 - 552) * u, z: (480 - 320) * u });
  p = hardwareThrow(s, 1, 255, distances);
  assert.equal(p.high, true);
  assert.equal(p.key, -2);
});
test("zapper targeting uses inclusive rectangle, strict aggression and skips electroballs", () => {
  const s = fixture(100, 900);
  let p = hardwareThrow(s, 1, 49, distances);
  assert.equal(p.high, false);
  assert.equal(p.key, -1);
  assert.equal(hardwareThrow(s, 1, 50, distances), null);
  s.ball.charged = true;
  assert.equal(hardwareThrow(s, 1, 49, distances), null);
  s.ball.charged = false;
  Object.assign(s.players[1], { x: (576 - 986) * u, z: (126 - 320) * u });
  assert.equal(hardwareThrow(s, 1, 49, distances).high, false);
  s.players[1].z = (127 - 320) * u;
  assert.equal(hardwareThrow(s, 1, 49, distances), null);
});
test("carrier starts hardware throw without legacy danger or goal-distance trigger", () => {
  const s = fixture(56, 700);
  step(s, simulationStep, {}, [false, false]);
  assert.equal(s.players[1].throwMode, 2);
  assert.equal(s.players[1].fx, 1);
  assert.equal(s.players[1].fz, 0);
});
